src/socket/mod.rs

```diff
@@ -0,0 +1,311 @@
+pub mod path;
+
+use std::convert::TryFrom;
+use std::io::Write;
+use std::marker::PhantomData;
+use std::os::unix::net::UnixStream;
+use std::time::{Duration, Instant};
+
+/// Wrapper around a socket that can send and receive structured messages.
+///
+/// `timeout` arguments set the socket timeout before reading/writing.
+pub struct ReadWriter<'a, R, W> {
+    // where R: serde::Deserialize {
+    socket: &'a UnixStream,
+    phantom_r: PhantomData<R>,
+    phantom_w: PhantomData<W>,
+}
+
+/// Milliseconds accepted by a `Timeout`.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct Millis(u16);
+
+impl From<Millis> for Duration {
+    fn from(m: Millis) -> Duration {
+        let Millis(u) = m;
+        Duration::from_millis(u64::from(u))
+    }
+}
+
+impl TryFrom<Duration> for Millis {
+    type Error = ();
+    fn try_from(d: Duration) -> Result<Millis, Self::Error> {
+        Ok(Millis(u16::try_from(d.as_millis()).map_err(|_| ())?))
+    }
+}
+
+/// A (possible) timeout.
+#[derive(Clone)]
+pub enum Timeout {
+    /// Do not time out.
+    Infinite,
+    /// Time out after `Duration`.
+    D(Millis),
+}
+
+impl Timeout {
+    /// Construct from a millisecond u16.
+    pub const fn from_millis(m: u16) -> Timeout {
+        Timeout::D(Millis(m))
+    }
+}
+
+impl std::fmt::Display for Timeout {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Timeout::Infinite => write!(f, "infinite timeout"),
+            Timeout::D(Millis(ms)) => write!(f, "{}ms", ms),
+        }
+    }
+}
+
+/// Reading from a `ReadWriter<R, W>` failed.
+#[derive(Debug)]
+pub enum ReadError {
+    /// Deserializing `R` failed.
+    Deserialize(bincode::Error),
+    /// No value available within given timeout.
+    Timeout,
+}
+
+// TODO: combine with ReadError?
+/// Writing to a `ReadWriter<R, W>` failed.
+#[derive(Debug)]
+pub enum WriteError {
+    /// Serializing `W` failed.
+    Serialize(bincode::Error),
+    /// No value available within given timeout.
+    Timeout,
+}
+
+impl From<bincode::Error> for WriteError {
+    fn from(e: bincode::Error) -> WriteError {
+        WriteError::Serialize(e)
+    }
+}
+
+/// Reading from or writing to a `ReadWriter<R, W>` failed.
+#[derive(Debug)]
+pub enum ReadWriteError {
+    /// Reading failed.
+    R(ReadError),
+    /// Writing failed.
+    W(WriteError),
+}
+
+impl From<ReadError> for ReadWriteError {
+    fn from(r: ReadError) -> Self {
+        ReadWriteError::R(r)
+    }
+}
+impl From<WriteError> for ReadWriteError {
+    fn from(r: WriteError) -> Self {
+        ReadWriteError::W(r)
+    }
+}
+
+// put into the Io part of a `bincode::ErrorKind`
+fn into_bincode_io_error<T>(res: std::io::Result<T>) -> bincode::Result<T> {
+    res.map_err(|e| Box::new(bincode::ErrorKind::Io(e)))
+}
+
+/// Run action with a timeout, return the remaining timeout
+/// or `Err` if there’s no time remaining.
+///
+/// The interface is not very “rusty”, but it’s only used internally.
```

Comment:
I don't love it, but okay ;)

Reply:
Yeah, it’s not great. But it fixes a couple of bugs that were in the original logic & deduplicates that logic between `react` and `communicate`.